### DISCUSSION TITLE
`engine.URL.__str__` should censor out password

### DISCUSSION THREAD

**Discussion by ya55en** [sqlalchemy/sqlalchemy#8547] [Ideas]
Similar to #2821, using `str(engine.URL())` in logs and prints can easily expose the database password to outside world as logs are usually stored into search engine backends to be available for all kinds of queries. I myself have recently caused such a leakage unintentionally using "%s" with a URL in a log statement, which brought lots of headaches to me and my colleagues.

A **"don't show until explicitly told to"** policy about passwords seems best IMHO. For those who really want to print their password (and frankly, I wonder why would someone need that), `URL.render_as_string(hide_password=False)` should suffice.

Please consider this patch - I believe it will do no harm to people already logging their database URL via `str()` and will save others (like me) tons of sorrows ;) Thanks!

```diff
diff --git a/lib/sqlalchemy/engine/url.py b/lib/sqlalchemy/engine/url.py
index 8d80cfd1c..b8cd50f4d 100644
--- a/lib/sqlalchemy/engine/url.py
+++ b/lib/sqlalchemy/engine/url.py
@@ -643,7 +643,7 @@ class URL(NamedTuple):
         return s
 
     def __str__(self) -> str:
-        return self.render_as_string(hide_password=False)
+        return self.render_as_string()
 
     def __repr__(self) -> str:
         return self.render_as_string()
```

**Comment by CaselIT** [comment]
Hi,

It seems that the current logic is in place since it was first implemented in https://github.com/sqlalchemy/sqlalchemy/commit/610684bb080095dcd8a2ca6cef1ff45787e4bdcf

I personally don't have a preference, but if we do this change I would instead remove `__str__` making python fallback on `__repr__`

**Reply by ya55en** [reply]
:+1: - yep, definitely better - didn't think of that. Thanks.

**Comment by zzzeek** [comment]
personally I dont know that someone logging their database URL directly and making the mistake that you did is any more common than people who are stringifying URLs and passing them to create_engine() to make a new engine (also a not great idea), so a change like this is likely very backwards incompatible for a lot of people.     that said, you appear to have actually made the mistake of putting a cleartext URL in a log file.   We can do it for 2.0 but we will likely get a lot of complaints.   if you can send a PR with tests that would be helpful.

**Reply by ya55en** [reply]
> you appear to have actually made the mistake of putting a cleartext URL in a log file

My bad, fully admitting 🙄

> a change like this is likely very backwards incompatible for a lot of people

Huh, bad news, didn't think of this ... an "improvement" should not create more trouble than it saves.

> if you can send a PR with tests that would be helpful

I can for sure, if this makes sense.

Thanks for caring. And for SqlAlchemy (I am a user since v.0.7).

**Reply by CaselIT** [reply]
Maybe we could add another shorthand method to render the url complete of password?
like `Url().pain_str()` or `Url().str()` or similar so we can provide an easier migration for users that use str(Url()) to renter also the password?

The change could be motivated by security improvement, make it less likely that url passwords are exposed to log or similar?

**Reply by ya55en** [reply]
Another approach that comes to mind: Introduce a keyword argument for constructing `URL()` objects, like `expose_pass` with a `False` default value. Everything else in the `URL` interface remains the same. Or at least old methods continue to work as now, with the exception that where the password is really desired in the `str(URL())` outcome, those users would need to add `expose_pass=True` to the `URL` creation statement ... Makes sense?

**Reply by zzzeek** [reply]
we have render_as_string() already, why have more than One Way to Do It ?

**Reply by zzzeek** [reply]
@ya55en im not sure if that would be too useful because you have to assume someone affected by this has like "str(engine.url)" like you do already, and they'd need to change their code regardless.  I think we can just drop `__str__()` and go with the change, people have been grumpy about URL changes lately but this one has "Security" written on it, also we'd get a chance to see if anyone is actually looking for URL stringify with round trips (we probably are doing so in our own test suite, but that's because we're testing URL itself).

**Reply by CaselIT** [reply]
> we have render_as_string() already, why have more than One Way to Do It ?

the reasoning was to provide a more convenient api to replace `str(Url())`. `Url().render_as_string(hide_password=False)` works but it's not too convenient. 
this was just a thought. we will probably not need it

**Reply by ya55en** [reply]
Okay, so I can remove `URL.__str__()` and try fixing the tests and submit a PR. IMHO, sounds like a good step towards making `URL` foolproof from security viewpoint.

**Reply by zzzeek** [reply]
OK this a thing I do which is I purposely leave things out of APIs to see what people ask for.  so in that sense `render_as_string(big_long_argment)` is in some ways an invitation to tell us what they are using it for.   also just switching password to be concealed here, while I'm curious about backwards incompat I am also looking to see what comes up.

**Reply by zzzeek** [reply]
like for example, all the new deferred docs I did yesterday, how come we have "load_only()" but not "defer_only()"?  things like that.    I assume w/ all the downloads we get, a lot of people are capable of asking for that feature if they actually need it.

**Comment by ya55en** [comment]
@CaselIT, @zzzeek: I ~~am ready to submit~~ submitted a PR (`main` branch), HTH.

Does it make sense to convert this discussion into an issue and refer to it in the commit message/the PR?
Thanks!

**Reply by zzzeek** [reply]
we can reference the discussion in the commit, I've been loose about that.

the PR needs a changelog file.  take a look in doc/build/changelog/unreleased_20/ to see examples.  tags for this one would be `bug, engine`.  thanks for your attentiveness!

**Reply by ya55en** [reply]
Done, used the discussion ID as a reference (and filename). Hope I got it right this time ... if not let me know; also I can squash the commits into one if that is preferred. Thank you!

**Comment by ya55en** [comment]
Thanks for considering this proposal and adopting the change. Thanks for SqlAlchemy!

**Comment by jfuruness** [comment]
Sorry to bring this up again but it looks like __str__ still exists, it was pretty confusing trying to figure out why the database connection was failing when passing in str(URL) (and it was because stringifying the URL redacts the password under the hood). That is really non-intuitive behavior to me. I understand the security concern, but perhaps a potential solution can be:

1. add two methods: stringify_w_password, stringify_wout_password
2. deprecate __str__ with a message saying to use one of the two methods above

I think this would make it a lot more intuitive for most users. For users logging the URI, they wouldn't be able to log it with the password accidentally, and for users just trying to connect to the database, their connections would not start having unexpected difficulties.

**Reply by CaselIT** [reply]
>I think this would make it a lot more intuitive for most users. For users logging the URI, they wouldn't be able to log it with the password accidentally, and for users just trying to connect to the database, their connections would not start having unexpected difficulties.

that's already how it's working. there is no need to use `str(URL)` to connect, if you already have the url just use it.

> add two methods: stringify_w_password, stringify_wout_password

already present, `render_as_string`

> deprecate str with a message saying to use one of the two methods above

I don't see how this would improve thins, also str(url) is a valid use case for log purposes

**Reply by jfuruness** [reply]
Certain applications like Alembic which integrate with sqlalchemy require a string, not a URL object. render_as_string imo does not explain to the user that it's not actually rendering the URL as a string, it is modifying it. I hear you when you say str(url) is a valid use case for log purposes, but to me it's also a valid use case for connection purposes, and str(URL) hides that it's not actually printing the string, it's redacting parts of it unbeknownst to the user.

If that's just the way it is no problem, but I think if you asked a random programmer what str(URL) would do, they would expect it to print the string of the URL and not modify parts of it. That's just my opinion though.

**Reply by zzzeek** [reply]
well str() would give you the full URL with the password for many years, then it found it's way into people's log file things, then we got a CVE.  So that's the way it is.  

I dont see how making str() raise an error helps anything.  if you write some code, and you have a str(url) in there and you try to use that to log into a database, it will fail.   because the password is all asterisks.  Why isn't that enough of a failure mode?

**Reply by CaselIT** [reply]
> Certain applications like Alembic which integrate with sqlalchemy require a string, not a URL object

that's when loaded from an ini file. but you don't have an url object there, you already start from a string. I'm not really sure were in alembic you need the string given the url.

> If that's just the way it is no problem, but I think if you asked a random programmer what str(URL) would do, they would expect it to print the string of the URL and not modify parts of it. That's just my opinion though.

I agree with you, and for a long time it was like that. We decided to change it for security purposes, so while inconvenient I think it was a valid change.

If adding a documentation note here https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.engine.URL to indicate this may help we can do that, but I think changing this again will just cause more confusion.

**Reply by zzzeek** [reply]
> render_as_string imo does not explain to the user that it's not actually rendering the URL as a string, it is modifying it.

What?  It actually renders the URL as a string.  if you know about the method, it means you read [the docstring](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.engine.URL.render_as_string) and IMO it's pretty hard to miss that big "hide_password=True" parameter right there, not to mention IDE will show the parameter, etc.

>  str(URL) hides that it's not actually printing the string, it's redacting parts of it unbeknownst to the user.

What?  just run it and look at the output, it's very obvious the password was obfuscated.  This is an common software development pattern for objects with security sensitive details to hide such details without special instructions

```py
>>> import psycopg2
>>> c = psycopg2.connect(user='scott', password='tiger', host="localhost", database="test")
>>> c
<connection object at 0x7f90961eda80; dsn: 'user=scott password=xxx dbname=test host=localhost', closed: 0>
```

**Reply by jfuruness** [reply]
Personally I think taking the string of a URL object and taking the string of a psycopg2 object aren't quite the same thing. URLs are used as strings as a parameter for connecting to databases. psycopg2 connection strings are not. If I print the string of a URL object, I'd expect to get back the URL. If I printed the string of a psycopg2 object, I'm not expecting to get back the URL of the database or even anything useful. 

As to why failing to connect isn't enough of a failure mode to realize that redaction is occurring:
If you pass in the string of a URL to form a connection to a postgresql database, it errors out and says the authentication failed, which led me down a rabbit hole of trying to figure out what I had misconfigured. Maybe a bad password, didn't whitelist the IP address, who knows. I also printed all the variables that I passed to the URL, and they were all correct. It turned out trying to get the string of the URL was modifying the contents under the hood, when I had just trusted that it was spitting out the URL, since after all it is a URL object.

I guess there was a change but I was just reflecting earlier, similar comments from zzzeek:
"@ya55en im not sure if that would be too useful because you have to assume someone affected by this has like "str(engine.url)" like you do already, and they'd need to change their code regardless. I think we can just drop __str__()"

while it was a long time ago, it also looks like others have encountered similar confusion: https://stackoverflow.com/q/59879387/8903959

It's odd to me that there's a security standard for redacting passwords in a URL under the hood when imo users shouldn't log strings containing passwords, but I guess that's life and there was a CVE for this and so it's not going to change. All good, I didn't realize that there was a CVE related to this issue. What is the CVE number btw? I couldn't find it online.

**Reply by CaselIT** [reply]
> URLs are used as strings as a parameter for connecting to databases

they shouldn't be used as strings

> If you pass in the string of a URL to form a connection to a postgresql database, it errors out and says the authentication failed, which led me down a rabbit hole of trying to figure out what I had misconfigured. Maybe a bad password, didn't whitelist the IP address, who knows. I also printed all the variables that I passed to the URL, and they were all correct. It turned out trying to get the string of the URL was modifying the contents under the hood, when I had just trusted that it was spitting out the URL, since after all it is a URL object.

It seems strange not to start by printing what you passed to create engine.

> It's odd to me that there's a security standard for redacting passwords in a URL under the hood when imo users shouldn't log strings containing passwords, but I guess that's life and there was a CVE for this and so it's not going to change. All good, I didn't realize that there was a CVE related to this issue. What is the CVE number btw? I couldn't find it online.

But that's not a password string, it's an object containing credentials. There is no cve, but it's usually how they get reported.

Overall you are complaining about a misunderstanding of what url is. It's an object that contains credentials not a password string. If the doc give the wrong impression could you indicate what paragraph should be improved?

Also is somewhere suggesting to use `create_engine(str(url))`? This is a misunderstanding of the API 

> while it was a long time ago, it also looks like others have encountered similar confusion: https://stackoverflow.com/q/59879387/8903959

This seems completely unrelated. They are printing an engine, why should it return the password? Why is it different that a psycog2 connection?

**Reply by jfuruness** [reply]
Definitely I do not understand the APIs at a deep level and the way this works is confusing to me, but perhaps that's just on me and I'm wrong. Honestly I don't think I will change your minds, and I don't think you will you change mine. As you both are (I assume) volunteers who have graciously dedicated your time to this project, I think lets agree to disagree. Thank you both for your explanations as well as your maintenance of this project, it is very well done, and I wish you both the best.

**Reply by CaselIT** [reply]
ok, that's fair. as mentioned I thinking doing further code changes will just compound to the confusion.

If you have documentation improvement that would have been helpful we are not opposed to that. Feel free to suggest them here or via a PR
